src/MongoDB.Driver.Core/Core/Authentication/Oidc/OidcInputConfiguration.cs

```
@@ -0,0 +1,98 @@
+            string principalName = null,
+            string providerName = null,
+            IRequestCallbackProvider requestCallbackProvider = null,
+            IRefreshCallbackProvider refreshCallbackProvider = null)
+        {
+            _endpoint = Ensure.IsNotNull(endpoint, nameof(endpoint));
+            _providerName = providerName; // can be null
+            _principalName = principalName; // can be null
+            _requestCallbackProvider = requestCallbackProvider; // can be null
+            _refreshCallbackProvider = refreshCallbackProvider; // can be null
+
+            EnsureOptionsValid();
+        }
+
+        public EndPoint EndPoint => _endpoint;
+        public bool IsCallbackWorkflow => _requestCallbackProvider != null || _refreshCallbackProvider != null;
+        public string PrincipalName => _principalName;
+        public string ProviderName => _providerName;
+        public IRequestCallbackProvider RequestCallbackProvider => _requestCallbackProvider;
+        public IRefreshCallbackProvider RefreshCallbackProvider => _refreshCallbackProvider;
+
+        public override int GetHashCode() =>
+            new Hasher()
+                .Hash(_providerName)
+                .Hash(_endpoint)
+                .Hash(_principalName)
+                .GetHashCode();
+
+        public override bool Equals(object obj)
+        {
+            if (object.ReferenceEquals(obj, null) || GetType() != obj.GetType()) { return false; }
+            var rhs = (OidcInputConfiguration)obj;
+            return
+                // This class is used as an oidc cache key for a callback workflow.
+                _providerName == rhs._providerName&&
+                EndPointHelper.Equals(_endpoint, rhs._endpoint) &&
+                object.Equals(_requestCallbackProvider, rhs._requestCallbackProvider) &&
+                object.Equals(_refreshCallbackProvider, rhs._refreshCallbackProvider) &&
+                _principalName == rhs._principalName; 
```

Remove extra space at the end.
nit: I would rearrange the comparisons, so simple comparisons will fail faster (leave the most complex to the end).